bpo-38891: avoid quadratic item access performance of ShareableList

tkren opened the pull request:
Avoid linear runtime of `ShareableList.__getitem__` and `ShareableList.__setitem__` by storing running allocated bytes in `ShareableList._allocated_bytes` instead of the number of bytes for a particular stored item.

<!-- issue-number: [bpo-38891](https://bugs.python.org/issue38891) -->
https://bugs.python.org/issue38891
<!-- /issue-number -->

tkren committed:
shared_memory: avoid quadratic item access performance of ShareableList

Avoid linear runtime of ShareableList.__getitem__ and
ShareableList.__setitem__ by storing running allocated bytes in
ShareableList._allocated_bytes instead of the number of bytes for
a particular stored item.

tkren committed:
add news entry

pitrou commented on Lib/multiprocessing/shared_memory.py:
```diff
@@ -293,9 +293,13 @@ def __init__(self, sequence=None, *, name=None):
             self._allocated_bytes = tuple(
-                    self._alignment if fmt[-1] != "s" else int(fmt[:-1])
-                    for fmt in _formats
+                (sum_allocated_bytes := sum_allocated_bytes + (
```
Hmm, please write it in a more readable way. For example:
```python
offset = 0
self._allocated_offsets = [0]
for fmt in _formats:
    offset += self._alignment if fmt[-1] != "s" else int(fmt[:-1])
    self._allocated_offsets.append(offset)
```

tkren replied:
@pitrou I rather not change the type of `self._allocated_bytes` to `list` when `sequence is not None`. Compare this to the call `self._allocated_bytes = struct.unpack_from(...)` when `sequence is None` in [line 352](https://github.com/tkren/cpython/blob/fb102fe7a981b3a7fd518a83714be274532d1046/Lib/multiprocessing/shared_memory.py#L352), which will create a `tuple` for `self._allocated_bytes`.

If you do not like the iterable-expression with the walrus operator as part of the tuple constructor call, we either

1. construct multiple tuples while looping over `_formats` (feels clumsy)
1. or sum over `_formats` similar to
   ```python
   self._allocated_bytes = tuple(
        itertools.accumulate(_formats,
                             func=lambda total, fmt: total + (self._alignment if fmt[-1] != "s" else int(fmt[:-1])),
                             initial=0)
    )
   ```
1. or use an inline function like
   ```python
   def _offsets():
      sum_allocated_bytes = 0
      for fmt in _formats:
         sum_allocated_bytes += self._alignment if fmt[-1] != "s" else int(fmt[:-1])
         yield sum_allocated_bytes

   self._allocated_bytes = tuple(offset for offset in _offsets())
   ```

Which option do you prefer? (Maybe there is a simpler way that I don't see now?)

pitrou replied:
Or just make it a list in all cases, which is the sanest thing to do IMHO.

pitrou reviewed (changes requested):
Thank you for spotting this issue. Just an improvement suggestion below.

tkren committed:
Improve readability

Always use a list for storing the sequence of running allocated bytes in
ShareableList._allocated_bytes

pitrou committed:
Improve readability, add comments
